### ISSUE TITLE
Performance regression

### ISSUE THREAD

**Issue by mikhail-mmlc** [microsoft/pyright#7813]
We are using Pyright v1.1.274 and want to upgrade to v1.1.360, but we've noticed a significant performance degradation: the execution time on our codebase went from 15s to 33s.

The error count is similar:
 - **v1.1.274**: 7729 errors
 - **v1.1.360**: 7957 errors

We checked the speeds of different versions of Pyright and found that **v1.1.328** nearly doubled in time on our codebase.

I've attached output from the `node --perf` command:
 - [pyright-1.1.274.perf.txt](https://github.com/microsoft/pyright/files/15172567/pyright-1.1.274.perf.txt)
 - [pyright-1.1.360.perf.txt](https://github.com/microsoft/pyright/files/15172569/pyright-1.1.360.perf.txt)

Let me know if there's anything else I can do to help narrow down what might be causing the slowdown.

<details><summary>List of versions with more than 10% difference in time</summary>

```
First entry: 1.1.274: 15.14s

1.1.292 -> 1.1.293: +19.13% 16.65s -> 19.83s
1.1.295 -> 1.1.296: -20.58% 20.97s -> 16.66s
1.1.311 -> 1.1.312: +10.45% 18.40s -> 20.33s
1.1.320 -> 1.1.321: -15.23% 21.94s -> 18.60s
1.1.327 -> 1.1.328: +99.64% 19.52s -> 38.97s
1.1.349 -> 1.1.350: -18.56% 40.19s -> 32.73s

Last entry: 1.1.360: 33.26s
```
</details>

<details><summary>List of all versions that we checked</summary>

```
Pyright Version, Execution Time (s), % Change
1.1.274:  15.14s   +0.00%
1.1.275:  15.23s   +0.62%
1.1.276:  15.81s   +3.83%
1.1.277:  14.69s   -7.13%
1.1.278:  14.72s   +0.26%
1.1.279:  15.21s   +3.27%
1.1.280:  15.13s   -0.52%
1.1.281:  15.49s   +2.39%
1.1.282:  15.13s   -2.32%
1.1.283:  15.20s   +0.49%
1.1.284:  14.94s   -1.74%
1.1.285:  14.99s   +0.32%
1.1.286:  15.12s   +0.89%
1.1.287:  16.29s   +7.77%
1.1.288:  16.11s   -1.13%
1.1.289:  16.54s   +2.67%
1.1.290:  16.45s   -0.52%
1.1.291:  16.25s   -1.21%
1.1.292:  16.65s   +2.43%
1.1.293:  19.83s  +19.13%
1.1.294:  20.24s   +2.07%
1.1.295:  20.97s   +3.61%
1.1.296:  16.66s  -20.58%
1.1.297:  16.23s   -2.56%
1.1.298:  16.55s   +1.94%
1.1.299:  16.22s   -1.96%
1.1.300:  17.14s   +5.68%
1.1.301:  16.72s   -2.47%
1.1.302:  17.81s   +6.53%
1.1.303:  17.76s   -0.30%
1.1.304:  17.91s   +0.87%
1.1.305:  17.92s   +0.04%
1.1.306:  18.08s   +0.92%
1.1.307:  17.88s   -1.15%
1.1.308:  18.00s   +0.69%
1.1.309:   5.17s   +0.00% - Failed
1.1.310:  18.40s   +2.25%
1.1.311:  18.40s   -0.00%
1.1.312:  20.33s  +10.45%
1.1.313:  20.71s   +1.88%
1.1.314:  22.64s   +9.32%
1.1.315:  21.47s   -5.15%
1.1.316:  21.59s   +0.52%
1.1.317:  22.06s   +2.21%
1.1.318:  21.54s   -2.36%
1.1.319:   0.00s   +0.00% - Failed (Installation error)
1.1.320:  21.94s   +1.83%
1.1.321:  18.60s  -15.23%
1.1.322:  19.06s   +2.52%
1.1.323:  19.18s   +0.60%
1.1.324:  19.08s   -0.51%
1.1.325:  19.05s   -0.15%
1.1.326:  19.42s   +1.94%
1.1.327:  19.52s   +0.51%
1.1.328:  38.97s  +99.64%
1.1.329:  41.10s   +5.46%
1.1.330:  40.58s   -1.25%
1.1.331:  39.49s   -2.70%
1.1.332:  38.32s   -2.96%
1.1.333:  39.13s   +2.11%
1.1.334:  40.76s   +4.16%
1.1.335:  41.26s   +1.22%
1.1.336:  40.44s   -1.98%
1.1.337:  40.79s   +0.87%
1.1.338:  41.12s   +0.79%
1.1.339:  40.45s   -1.61%
1.1.340:  39.60s   -2.10%
1.1.341:  41.24s   +4.14%
1.1.342:  41.41s   +0.41%
1.1.343:  41.72s   +0.75%
1.1.344:  41.63s   -0.22%
1.1.345:  40.11s   -3.65%
1.1.346:  40.22s   +0.27%
1.1.347:  40.24s   +0.05%
1.1.348:  39.90s   -0.83%
1.1.349:  40.19s   +0.72%
1.1.350:  32.73s  -18.56%
1.1.351:  33.57s   +2.55%
1.1.352:  32.94s   -1.87%
1.1.353:  32.63s   -0.92%
1.1.354:  32.70s   +0.20%
1.1.355:  32.91s   +0.64%
1.1.356:  32.81s   -0.31%
1.1.357:  33.45s   +1.96%
1.1.358:  33.10s   -1.04%
1.1.359:  33.40s   +0.91%
1.1.360:  33.26s   -0.44%
```
</details>

**Comment by erictraut** [comment]
Wow, you're still using version 274? That's quite outdated. Version 328 is pretty old too. This is going to be an archeological dig at this point. I realize there's a cost to staying up to date with your tools, but it would be helpful in the future if you try to stay more current.

Is your source base publicly available? If not, are you be willing to give me temporary access to your source base?Performance issues are very difficult to diagnose without source access. Without access, it will require significantly more work on your part and mine, and our odds of finding the root cause goes down significantly.

One thing I can suggest is that you run `pyright --stats --verbose`. This will output a list of all checked source files listed in order of analysis time. Sometimes, there will be a single file that sticks out as an obvious outlier. You can then go into that file and start to comment out part of the code, effectively binary searching for the class or function that's causing the slow-down. If you can then create a minimal self-contained repro that obfuscates the code, I can investigate more closely.

Looking back at the [release notes for 1.1.328](https://github.com/microsoft/pyright/releases/tag/1.1.328), I don't see anything that jumps out at me as an obvious suspect for a performance degradation.

**Comment by mikhail-mmlc** [comment]
The code is not publicly available, and I cannot provide even temporary access to it.
I'll try to help with diagnosing from my side.
`pyright --stats --verbose` is helpful; I've identified several files where there is a significant difference in timing between versions 1.1.274 and 1.1.360, and I've managed to extract a snippet from one of them.

```python
DETAILS = (('DV', 'IGD', '838', ('IGD',)), ('JS', 'PXX', '362', ('PXX',)), ... )  # > 200 items

def get_row(extra_info):
    # > 10 variables
    var1 = [v[3] for v in DETAILS if v[2] == extra_info.code]
    var2 = [v[3] for v in DETAILS if v[2] == extra_info.code]         
```

With the attached [example.py](https://github.com/microsoft/pyright/files/15173206/example.txt) I got:
 - v1.1.274: `346ms`
  - v1.1.360: `6121ms`

**Comment by erictraut** [comment]
Thanks, I'm able to repro the issue. I've narrowed it down to [this commit](https://github.com/microsoft/pyright/commit/4f74921d27490a793fa441e91762d8f9b32c0023) which was a fix for [this bug](https://github.com/microsoft/pyright/issues/5968).

I've implemented a fix, and it will be included in the next release.

If you're looking for a workaround in the meantime, you can add a type annotation to `DETAILS`.

```python
DETAILS: tuple[tuple[str, str, str, tuple[str]], ...] = (...)
```

**Comment by mikhail-mmlc** [comment]
Great! I'll check again once the new release is available to see if there are other outliers.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.362, which I just published.
